src/packagedcode/opam.py

```diff
@@ -0,0 +1,309 @@
+# CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+#
+# When you publish or redistribute any data created with ScanCode or any ScanCode
+# derivative work, you must accompany this data with the following acknowledgment:
+#
+#  Generated with ScanCode and provided on an "AS IS" BASIS, WITHOUT WARRANTIES
+#  OR CONDITIONS OF ANY KIND, either express or implied. No content created from
+#  ScanCode should be considered or used as legal advice. Consult an Attorney
+#  for any legal advice.
+#  ScanCode is a free software code scanning tool from nexB Inc. and others.
+#  Visit https://github.com/nexB/scancode-toolkit/ for support and download.
+
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import unicode_literals
+
+from collections import OrderedDict
+import io
+import logging
+import re
+
+import attr
+from packageurl import PackageURL
+
+from commoncode import filetype
+from commoncode import fileutils
+from packagedcode import models
+
+
+"""
+Handle opam package.
+"""
+
+TRACE = False
+
+logger = logging.getLogger(__name__)
+
+if TRACE:
+    import sys
+    logging.basicConfig(stream=sys.stdout)
+    logger.setLevel(logging.DEBUG)
+
+
+@attr.s()
+class OpamPackage(models.Package):
+    metafiles = ('*opam',)
+    extensions = ('.opam',)
+    default_type = 'opam'
+    default_primary_language = 'Ocaml'
+    default_web_baseurl = 'https://opam.ocaml.org/packages'
+    default_download_baseurl = None
+    default_api_baseurl = None
+
+    @classmethod
+    def recognize(cls, location):
+        yield parse(location)
+
+    @classmethod
+    def get_package_root(cls, manifest_resource, codebase):
+        return manifest_resource.parent(codebase)
+
+    def repository_homepage_url(self, baseurl=default_web_baseurl):
+        if self.name:
+            return '{}/{}'.format(baseurl, self.name)
+
+
+def is_opam(location):
+    if location.endswith('opam'):
+        return True
+
+
+def parse(location):
+    """
+    Return a Package object from a opam file or None.
+    """
+    if not is_opam(location):
+        return
+
+    package_data = parse_opam(location)
+    return build_opam_package(package_data)
+
+
+def build_opam_package(package_data):
+    """
+    Return a Package from a opam file or None.
+    """
+    package_dependencies = []
+    deps = package_data.get('depends') or []
+    for dep in deps:
+        package_dependencies.append(
+            models.DependentPackage(
+                purl=dep.purl,
+                requirement=dep.version,
+                scope='dependency',
+                is_runtime=True,
+                is_optional=False,
+                is_resolved=False,
+            )
+        )
+
+    name = package_data.get('name')
+    maintainers = package_data.get('maintainer') or []
+    authors = package_data.get('authors') or []
```

Comment:
can you move these down closer to the loop they are processed in?

Reply:
:+1: